text/0000-router-service.md

```diff
@@ -0,0 +1,439 @@
+
+### Example: testing whether route will remain active
+
+Here's an example showing how `willTransition` can figure out if the current route will remain active after the transition:
+
+```js
+willTransition(transition) {
+  if (!this.transition.to.find(route => route.name === this.routeName)) {
+    alert("Please save or cancel your changes.");
+    transition.abort();
+  }
+}
+```
+
+### Example: parent redirecting to a fallback model
+
+Here's an example of a parent route that can redirect to a fallback model, without losing its child route:
+
+```js
+this.route('person', { path: '/person/:person_id' }, function() {
+  this.route('index');
+  this.route('detail');
+});
+
+//PersonRoute
+const fallbackPersonId = 0;
+model({ personId }, transition) {
+  return this.get('store').find('person', personId).catch(err => {
+    this.replaceWith(transition.to.name, fallbackPersonId);
+  });
+}
+
+// If personId 5 is invalid, and the user visits /person/5/detail, they will get
+// redirected to /person/0/detail. And /person/5 will get redirected to /person/0.
+```
+
+
+### Actively discourage use of private API
+
+This RFC provides public API for doing the things people have become accustomed to doing via private API. To eliminate confusion over the correct way, we should hide all the private API away behind symbols, and provide deprecation warnings per our usual release policy around breaking "widely-used private APIs".
+
+Some of the private APIs we should mark and warn include:
+
+ - transition.state
+ - transition.params
+ - `lookup('router:main')` (should use `service:router` instead)
+
+
+## Dynamically-Scoped Variables
+
+The next section introduces general-purpose, dynamically-scoped variables. Why am I doing this in "The Router Service RFC"? Because dynamically-scoped RouteInfo has several nice capabilities that will be introduced in the subsequent section. And once we have one dynamically-scoped variable, we have the option of exposing that capability for other uses.
+
+### In the general case
+
+A [dynamically-scoped variable](https://en.wikipedia.org/wiki/Scope_(computer_science)#Dynamic_scoping) takes its value from its calling context, as opposed to its lexical context. Since dynamically-scoped variables are powerful to the point of potential danger, their syntax is intended to be appropriately verbose.
+
+Reading a dynamically-scoped variable in handlebars:
+
+ ```hbs
+ {{!- retrieve the value of a dynamically scoped variable }}
+ (get-dynamic-variable "myVariableName")
+ ```
+
+Defining a component that reads a dynamically-scoped variable:
+
+```js
+let MyComponent = Ember.Component.extend({
+  didInsertElement() {
+    // Access to the variable here is intended to be indistinguishable
+    // from a normal, explicitly-passed input argument. 
+    doSomethingWith(this.get('myDynamicVariable'));
+  }
+});
+MyComponent.reopenClass({
+  // This is where the specialness happens.
+  getDynamicVariables: [ 'myDynamicVariable' ]
```

Comment:
I considered requiring doing it in a template, but that has poor ergonomics. We want `get-dynamic-variable` to be an internal implementation detail, not something that callers need to know about. So you want to hide it inside the implementation of your helper or component.

You can do that in a component by nesting another private component inside it, and that is merely awkward:

```hbs
{{-my-inner-component info=(get-dynamic-variable "routeInfo")}}
```

But helpers are worse because there's no way to implement one in terms of hbs. 

As for the detailed semantic questions: the answer to all of them is that these paramters should be indistinguishable from parameters that are passed explicitly in the template. In theory, you could implement them with an AST transform that rewrites:

```hbs
{{my-component foo=bar}}
```

into

```hbs
{{my-component foo=bar myDynamicVariable=(get-dynamic-variable "myDynamicVariable")}}
```

I deliberately didn't specify any different semantics because they should always follow what other normal parameters do -- and that will differ between classic and glimmer components.